src/transformers/models/detr/modeling_detr.py

```diff
@@ -0,0 +1,2201 @@
+
+
+@dataclass
+class DetrForSegmentationOutput(DetrObjectDetectionOutput):
+    """
+    This class adds one attribute to DetrObjectDetectionOutput, namely predicted masks.
+
+    Args:
+        pred_masks (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, num_queries, width, height)`):
+            ...
+    """
+
+    pred_masks: torch.FloatTensor = None
+
+
+# BELOW: utilities copied from
+# https://github.com/facebookresearch/detr/blob/master/backbone.py
+class FrozenBatchNorm2d(nn.Module):
+    """
+    BatchNorm2d where the batch statistics and the affine parameters are fixed.
+
+    Copy-paste from torchvision.misc.ops with added eps before rqsrt, without which any other models than
+    torchvision.models.resnet[18,34,50,101] produce nans.
+    """
+
+    def __init__(self, n):
+        super(FrozenBatchNorm2d, self).__init__()
+        self.register_buffer("weight", torch.ones(n))
+        self.register_buffer("bias", torch.zeros(n))
+        self.register_buffer("running_mean", torch.zeros(n))
+        self.register_buffer("running_var", torch.ones(n))
+
+    def _load_from_state_dict(
+        self, state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs
+    ):
+        num_batches_tracked_key = prefix + "num_batches_tracked"
+        if num_batches_tracked_key in state_dict:
+            del state_dict[num_batches_tracked_key]
+
+        super(FrozenBatchNorm2d, self)._load_from_state_dict(
+            state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs
+        )
+
+    def forward(self, x):
+        # move reshapes to the beginning
+        # to make it user-friendly
+        w = self.weight.reshape(1, -1, 1, 1)
+        b = self.bias.reshape(1, -1, 1, 1)
+        rv = self.running_var.reshape(1, -1, 1, 1)
+        rm = self.running_mean.reshape(1, -1, 1, 1)
+        eps = 1e-5
+        scale = w * (rv + eps).rsqrt()
+        bias = b - rm * scale
+        return x * scale + bias
+
+
+class Backbone(nn.Module):
+    """
+    Timm convolutional backbone.
+
+    If resnet, replace nn.BatchNorm2d layers by FrozenBatchNorm2d defined above.
+
+    """
+
+    def __init__(self, name: str, train_backbone: bool, dilation: bool, return_intermediate_layers: bool):
+        super().__init__()
+
+        kwargs = {}
+        if name in ["resnet18", "resnet34", "resnet50", "resnet101"]:
+            kwargs["norm_layer"] = FrozenBatchNorm2d
+        if dilation:
+            kwargs["output_stride"] = 16
+        self.body = create_model(name, pretrained=True, features_only=True, **kwargs)
+        self.return_intermediate_layers = return_intermediate_layers
+        self.num_channels = self.body.feature_info.channels(-1)
+
+        for name, parameter in self.body.named_parameters():
+            if not train_backbone or "layer2" not in name and "layer3" not in name and "layer4" not in name:
+                parameter.requires_grad_(False)
```

Comment:
There should be a method on the model to freeze/unfreeze that backbone. It can be frozen by default if this is the better choice, but it should not be parametrized by the config.

Reply:
Ok, I'll remove it, although it may be handy for users, no? cc @LysandreJik